libs/sdk-ui-pivot/src/impl/resizing/columnSizing.ts

```diff
@@ -734,87 +738,103 @@ function valueFormatter(text: string, col: SeriesCol, separators: any) {
-            }
+    const col = config.tableDescriptor.getCol(column);
+
+    if (col && context) {
+        const text = row[col.id];
+        const formattedText = isSeriesCol(col) && valueFormatter(text, col, config.separators);
+        const textForCalculation = formattedText || text;
+        const maxWidth = col.id ? maxWidths.get(col.id) : undefined;
+        let possibleMaxWidth;
+
+        if (config.cache) {
+            possibleMaxWidth = getMaxWidthCached(context, textForCalculation, maxWidth, config.cache);
+        } else {
+            possibleMaxWidth = getMaxWidth(context, textForCalculation, false, maxWidth);
         }
-    });
+
+        if (possibleMaxWidth) {
+            maxWidths.set(col.id, possibleMaxWidth);
+        }
+    }
+}
+
+function getColWidth(maxWidth: number | undefined, padding: number): number {
+    const newWidth = maxWidth ? Math.ceil(maxWidth + padding) : 0;
+    return Math.min(Math.max(MIN_WIDTH, newWidth), AUTO_SIZED_MAX_WIDTH);
 }
 
 export function getUpdatedColumnDefs(
     columns: Column[],
     maxWidths: Map<string, number>,
     padding: number,
 ): ColDef[] {
-    return columns.map((column: Column) => {
+    return columns.reduce((updatedColumnDefs: ColDef[], column: Column) => {
         const colDef: ColDef = column.getColDef();
         const colId = agColId(colDef);
 
         if (colId) {
             const maxWidth = maxWidths.get(colId);
-            const newWidth = maxWidth ? Math.ceil(maxWidth + padding) : 0;
-
-            return {
-                ...colDef,
-                width: Math.min(Math.max(MIN_WIDTH, newWidth), AUTO_SIZED_MAX_WIDTH),
-            };
+            if (maxWidth) {
+                updatedColumnDefs.push({
+                    ...colDef,
+                    width: getColWidth(maxWidth, padding),
+                });
+            }
         }
 
-        return colDef;
-    });
+        return updatedColumnDefs;
+    }, []);
 }
 
+const shouldStopCalculation = (config: CalculateColumnWidthsConfig, calculatedColumnsTotalWidth: number) =>
+    config.columnAutoresizeOption === "viewport" &&
+    config.clientWidth &&
+    calculatedColumnsTotalWidth > config.clientWidth;
+
 function calculateColumnWidths(config: CalculateColumnWidthsConfig) {
     const { context } = config;
     const maxWidths = new Map<string, number>();
-
-    if (config.measureHeaders && context) {
-        context.font = config.headerFont;
-
+    if (context) {
+        let calculatedColumnsTotalWidth = 0;
         config.columns.forEach((column: Column) => {
             const colDef: ColDef = column.getColDef();
             const colId = agColId(colDef);
             const maxWidth = colId ? maxWidths.get(colId) : undefined;
-            const possibleMaxWidth = getMaxWidth(context, colDef.headerName, !!colDef.sort, maxWidth);
+            if (shouldStopCalculation(config, calculatedColumnsTotalWidth)) {
+                return;
```

Comment:
Nitpick: forEach can be probably replaced with for cycle, so it's not necessary to iterate through the rest of the columns after reaching clientWidth